tests/protocol/test_wire.py

```diff
@@ -566,6 +563,176 @@ def test_report_event_large_event(self, patch_send_event, *args):
+        xml_text = WireProtocolData(DATA_FILE).goal_state
+        client = WireClient(wireserver_url)
+        client.goal_state = GoalState(xml_text)
+        client.get_host_plugin().set_default_channel(True)
+
+        def direct_func(*args):
+            direct_func.counter += 1
+            return 42
+
+        def host_func(*args):
+            host_func.counter += 1
+            return 43
+
+        direct_func.counter = 0
+        host_func.counter = 0
+
+        # Assert we've only called the host channel function since it's the default channel
+        ret = client.send_request_using_appropriate_channel(direct_func, host_func)
+        self.assertEquals(43, ret)
+        self.assertEquals(0, direct_func.counter)
+        self.assertEquals(1, host_func.counter)
+
+    def test_send_request_using_appropriate_channel_direct_func_fails_host_func_works_first_time(self, *args):
+        xml_text = WireProtocolData(DATA_FILE).goal_state
+        client = WireClient(wireserver_url)
+        client.goal_state = GoalState(xml_text)
+        host = client.get_host_plugin()
+        host.set_default_channel(False)
+
+        def direct_func(*args):
+            direct_func.counter += 1
+            raise InvalidContainerError()
+
+        def host_func(*args):
+            host_func.counter += 1
+            return 42
+
+        direct_func.counter = 0
+        host_func.counter = 0
+
+        # Assert we've called both the direct channel function and the host channel function, which succeeded.
+        # After the host channel succeeds, the host plugin should have been set as the default channel.
+        ret = client.send_request_using_appropriate_channel(direct_func, host_func)
+        self.assertEquals(42, ret)
+        self.assertEquals(1, direct_func.counter)
+        self.assertEquals(1, host_func.counter)
+        self.assertEquals(True, host.is_default_channel())
+
+    def test_send_request_using_appropriate_channel_direct_func_fails_host_func_succeeds_second_time(self, *args):
+        xml_text = WireProtocolData(DATA_FILE).goal_state
+        client = WireClient(wireserver_url)
+        client.goal_state = GoalState(xml_text)
+        client.get_host_plugin().set_default_channel(False)
+
+        def direct_func(*args):
+            direct_func.counter += 1
+            raise InvalidContainerError()
+
+        def host_func(*args):
+            host_func.counter += 1
+            if host_func.counter == 1:
+                raise ResourceGoneError("Resource is gone")
+            return 42
+
+        direct_func.counter = 0
+        host_func.counter = 0
+
+        # Assert we've called both the direct channel function (once) and the host channel function (twice).
+        # After the host channel succeeds, the host plugin should have been set as the default channel.
+        with patch('azurelinuxagent.common.protocol.wire.WireClient.update_goal_state'):
```

Comment:
should also assert that the goal state was refreshed

Reply:
Thanks, added.